File: Java/src/com/interpreter/box/BoxFunction.java
package com.interpreter.box;

import java.util.List;                    

class BoxFunction implements BoxCallable {
  private final Stmt.Function declaration;
  private final Environment closure;
  private final boolean isInitializer;

  BoxFunction(Stmt.Function declaration, Environment closure,
              boolean isInitializer) {                       
    this.isInitializer = isInitializer;
	    this.closure = closure;
    this.declaration = declaration;       
  }     
  
  BoxFunction bind(BoxInstance instance) {             
	    Environment environment = new Environment(closure);
	    environment.define("this", instance);              
	    return new BoxFunction(declaration, environment, isInitializer); 
	  }
  
  @Override                                       
  public String toString() {                      
    return "<fn " + declaration.name.lexeme + ">";
  }
  
  @Override                          
  public int arity() {               
    return declaration.params.size();
  }
  
  @Override                                                            
  public Object call(Interpreter interpreter, List<Object> arguments) {
	  Environment environment = new Environment(closure);    
    for (int i = 0; i < declaration.params.size(); i++) {              
      environment.define(declaration.params.get(i).lexeme,             
          arguments.get(i));                                           
    }

    try {                                                     
        interpreter.executeBlock(declaration.body, environment);
      } catch (Return returnValue) { 
    	  if (isInitializer) return closure.getAt(0, "this");
        return returnValue.value;                               
      }
    
    return null;                                                       
  }
}

File: C#/Interpreter/src/Box/BoxFunction.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Interpreter.Box
{
    public class BoxFunction : BoxCallable
    {
        private Stmt.Function declaration;
        private Environment closure;
        private bool isInitializer;

        public BoxFunction(Stmt.Function declaration, Environment closure, bool isInitializer)
        {
            this.isInitializer = isInitializer;
            this.closure = closure;
            this.declaration = declaration;
        }

        public BoxFunction bind(BoxInstance instance)
        {
            Environment environment = new Environment(closure);
            environment.define("this", instance);
            return new BoxFunction(declaration, environment, isInitializer);
        }

        public string toString()
        {
            return "<fn " + declaration.name.lexeme + ">";
        }

        public int arity()
        {
            return declaration.parameters.Count;
        }

        public object call(Interpreter interpreter, List<object> arguments)
        {
            Environment environment = new Environment(closure);
            for (int i = 0; i < declaration.parameters.Count; i++) {
                environment.define(declaration.parameters[i].lexeme, arguments[i]);
            }

            try
            {
                interpreter.executeBlock(declaration.body, environment);
            }
            catch (Return returnValue)
            {
                if (isInitializer) return closure.getAt(0, "this");
                return returnValue.value;
            }

            if (isInitializer) return closure.getAt(0, "this");

            return null;
        }
    }
}
